### DISCUSSION TITLE
Is there a way to copy PowerShell buffer using one of Windows Terminal actions?

### DISCUSSION THREAD

**Discussion by gaberiel44** [microsoft/terminal#15349] [Q&A] [answered]
I am in a bit of a dilemma here, I will explain myself first.

In Windows Terminal I only ever use PowerShell (7.3) and in my PowerShell profile I have the following:
```
Set-PSReadLineKeyHandler -Chord Ctrl+c -Function Copy
```

I am trying to get Copy to clipboard on the terminal to conform to way it works everywhere else on Windows: Only copy to clipboard IF there is a selection

I cant seem to find a way around:  How **Copy to Clipboard** is handled on the terminal. 
In Notepad, if there is a selection, pressing `Ctrl+C`, the text will be copied to the clipboard, otherwise nothing happens. This is the same in other edit boxes and controls

In Windows Terminal (More specifically, PowerShell)  this is not so. 
While there is a selection pressing `Ctrl+C`, the text will be copied to the clipboard, otherwise the buffer is copied to the clipboard.
To make things worse, after pressing `Ctrl` and `C`  the users selection  is discarded.

The following is a recording from a terminal window, when I copy the text, the selection is discarded:
![](https://i.imgur.com/z0EJ68M.gif)

In this recording from Notepad, when I copy the selection to the clipboard, the selection is maintained:
![](https://i.imgur.com/sZmuBT0.gif)

I would like Terminal/PowerShell to behave exactly like Notepad

From this article [Selecting text in Windows Terminal](https://learn.microsoft.com/en-us/windows/terminal/selection) part of Windows Terminals documentation:
> Once you have a selection present, you have a few options...you can right-click to copy the selected text to your clipboard and clear the selection. If you right-click again, the contents of your clipboard will then be pasted into the terminal.

This would be a decent solution (I can assign a keyboard key to simulate a right click) but the above does not apply to the buffer (I mean, the text currently being typed) only to the terminals previous output.

On this page [Clipboard integration commands | Windows Terminal Actions](https://learn.microsoft.com/en-us/windows/terminal/customize-settings/actions#clipboard-integration-commands)
>Copy
This copies the selected terminal content to your clipboard. If no selection exists, the key chord is sent directly to the terminal.

"This copies the selected terminal content to your clipboard" I think "terminal content" is meant as the terminal output, NOT the buffer. Is there a way to get the `copy` action to also copy the terminals buffer rather than passing the key directly to the terminal? 

VSCode by default, has similar behaviour, but offers a means to disable it:
```
"editor.emptySelectionClipboard": false   //<=== With this, pressing Ctrl + C without a selection, does nothing.
```

I am asking for a specific terminal method because looking at the all the [PSReadLine class methods](https://learn.microsoft.com/en-us/dotnet/api/microsoft.powershell.psconsolereadline?view=powershellsdk-1.1.0), no solution is provided for just performing conventional "copy to clipboard", these are the only two copy methods available:
```
Copy(Nullable<ConsoleKeyInfo>, Object)
Copy selected region to the system clipboard. If no region is selected, copy the whole line.

CopyOrCancelLine(Nullable<ConsoleKeyInfo>, Object)
If text is selected, copy to the clipboard, otherwise cancel the line.
```

Thanks for any help.

**Comment by zadjii-msft** [comment, marked as answer]
> "This copies the selected terminal content to your clipboard" I think "terminal content" is meant as the terminal output, NOT the buffer.

Here's a trick: the terminal doesn't actually know the difference between "output" and "input" in the buffer :wink:
![dcbc6e11-44f2-492e-8bb9-e124601f31ba](https://github.com/microsoft/terminal/assets/18356694/9971f594-e964-4162-8df6-47bceef08afa)


So that's me using Windows Terminal's built-in <kbd>ctrl+c</kbd> "Copy" action. If there's a selection, it'll copy and dismiss the selection. If there's no selection, it'll just send a `^C` to the connected client app. 

> In Notepad, if there is a selection, pressing Ctrl+C, the text will be copied to the clipboard, otherwise nothing happens. This is the same in other edit boxes and controls
> ... 
> I would like Terminal/PowerShell to behave exactly like Notepad

This seems like a reasonable property to add to the `copy` action. 

```ts
class CopyActionArgs {
  singleLine: bool = false;
  copyFormatting = false;
  dismissSelection: bool = false;
}
```

Seems like there somehow isn't a thread for this yet! Lemme go make one.

**Reply by zadjii-msft** [reply]
I've moved this request to #15371
